File: solutions/opus-1/opus/java/anagrams.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        Map<String, List<String>> groups = new HashMap<>();
        String line;
        while ((line = in.readLine()) != null) {
            String w = line.trim();
            if (w.isEmpty()) continue;
            char[] c = w.toCharArray();
            Arrays.sort(c);
            groups.computeIfAbsent(new String(c), k -> new ArrayList<>()).add(w);
        }
        List<List<String>> list = new ArrayList<>(groups.values());
        for (List<String> g : list) Collections.sort(g);
        list.sort((a, b) -> a.get(0).compareTo(b.get(0)));
        StringBuilder sb = new StringBuilder();
        for (List<String> g : list) sb.append(String.join(" ", g)).append('\n');
        System.out.print(sb);
    }
}

File: solutions/opus-1/opus/rust/anagrams.rs
use std::collections::HashMap;
use std::io::{self, Read, Write};

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut groups: HashMap<Vec<u8>, Vec<&str>> = HashMap::new();
    for line in input.lines() {
        let w = line.trim();
        if w.is_empty() {
            continue;
        }
        let mut key = w.as_bytes().to_vec();
        key.sort_unstable();
        groups.entry(key).or_default().push(w);
    }
    let mut out: Vec<Vec<&str>> = groups.into_values().collect();
    for g in out.iter_mut() {
        g.sort_unstable();
    }
    out.sort_unstable_by(|a, b| a[0].cmp(b[0]));
    let stdout = io::stdout();
    let mut o = io::BufWriter::new(stdout.lock());
    for g in out {
        writeln!(o, "{}", g.join(" ")).unwrap();
    }
}
